adapter/src/main/java/org/jvalue/ods/adapterservice/datasource/api/amqp/AmqpPublisher.java

```diff
@@ -0,0 +1,61 @@
+package org.jvalue.ods.adapterservice.datasource.api.amqp;
+
+import org.jvalue.ods.adapterservice.config.RabbitConfiguration;
+import org.jvalue.ods.adapterservice.datasource.event.*;
+import org.jvalue.ods.adapterservice.datasource.model.Datasource;
+import org.springframework.amqp.AmqpException;
+import org.springframework.amqp.rabbit.core.RabbitTemplate;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+import java.io.Serializable;
+
+@Service
+public class AmqpPublisher {
+
+    private final RabbitTemplate rabbitTemplate;
+
+    @Autowired
+    public AmqpPublisher(RabbitTemplate rabbitTemplate) {
+        this.rabbitTemplate = rabbitTemplate;
+    }
+
+    public void publishCreation(Datasource datasource) {
+        publish(RabbitConfiguration.AMQP_DATASOURCE_CREATED_TOPIC, new DatasourceCreationEvent(datasource));
+    }
+
+    public void publishUpdate(Datasource datasource) {
+        publish(RabbitConfiguration.AMQP_DATASOURCE_UPDATED_TOPIC, new DatasourceUpdateEvent(datasource));
+    }
+
+    public void publishDeletion(Long datasourceId) {
+        publish(RabbitConfiguration.AMQP_DATASOURCE_DELETED_TOPIC, new DatasourceConfigEvent(datasourceId));
+    }
+
+    public void publishImportSuccess(Long id, String data) {
+        publish(RabbitConfiguration.AMQP_IMPORT_SUCCESS_TOPIC, new DatasourceImportedEvent(id, data));
+    }
+
+    public void publishImportFailure(Long id, String errMsg) {
+        publish(RabbitConfiguration.AMQP_IMPORT_FAILED_TOPIC, new ImportFailedEvent(id, errMsg));
+    }
+
+    private void publish(String topic, Serializable message) {
+        for (int retries = RabbitConfiguration.AMQP_PUBLISH_RETRIES; retries >= 0; retries--) {
+            try {
+                this.rabbitTemplate.convertAndSend(RabbitConfiguration.AMPQ_EXCHANGE, topic, message);
+                return;
+            } catch (AmqpException e) {
+                try {
+                    Thread.sleep(RabbitConfiguration.AMQP_PUBLISH_BACKOFF);
+                } catch (InterruptedException interruptedException) {
+                    Thread.currentThread().interrupt();
+                    throw new RuntimeException(interruptedException);
+                }
+                System.out.println("Message publish failed ("+retries+"). Retrying in "+RabbitConfiguration.AMQP_PUBLISH_BACKOFF);
```

Comment:
Only log in debug mode?

Reply:
You mean using system.err or using the spring logging framework?

Reply:
No, just thinking in general how to handle that error... Printing seems to be a good way since we do retries. There, we have a few log levels in general (debug, info, warning, error) where we can indicate stuff. Probably we want do define when to use which log levels in the future. Nvm for now.